// Copyright 2021 TsumiNa
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
All files should be ordered in the following manner:

1. License header (if applicable)
2. File overview
3. Syntax
4. Package
5. Imports (sorted)
6. File options
7. Everything else

see also: https://developers.google.com/protocol-buffers/docs/style
*/

syntax = "proto3";

option go_package = "github.com/yoshida-lab/avalon/protobuf/scheduler";
package scheduler;

import "proto/task/task.proto";

service Scheduler {
  rpc InitFrom(task.TaskRoot) returns (task.Task);
  rpc StepForward(SchedulerForwardReq) returns (SchedulerForwardRes);
}

enum SchedulerForwardFlag {
  FORWARD = 0;
  FINISH = 1;
  TERMINATE = 2;
}

message SchedulerSubmission {
  string job_dir = 1;
  string hostname = 2;
  map<string, string> overwrite = 3;
}

message SchedulerForwardReq {
  uint32 task_id = 1;
  SchedulerForwardFlag flag = 2;
  map<string, string> variables = 3;
  repeated SchedulerSubmission submissions = 4;
}

message SchedulerForwardRes {
  uint32 task_id = 1;
  SchedulerForwardFlag flag = 2;
  repeated uint32 job_ids = 3 [ packed = true ];
}
